packages/viewer/src/controllerModel.js

```diff
@@ -0,0 +1,159 @@
+/* eslint-enable */
+
+import AssetError from './assetError.js';
+
+const gltfLoader = new GLTFLoader();
+
+class ControllerModel extends THREE.Object3D {
+  constructor() {
+    super();
+    this.xrInputSource = null;
+    this.rootNode = null;
+    this.nodes = [];
+    this.motionController = null;
+    this.loaded = false;
+  }
+
+  async initialize(motionController) {
+    this.motionController = motionController;
+    this.xrInputSource = this.motionController.xrInputSource;
+
+    // Fetch the assets and generate threejs objects for it
+    this.asset = await new Promise(((resolve, reject) => {
+      gltfLoader.load(
+        motionController.assetUrl,
+        (loadedAsset) => { resolve(loadedAsset); },
+        null,
+        () => { reject(new AssetError(`Asset ${motionController.assetUrl} missing or malformed.`)); }
+      );
+    }));
+
+    this.rootNode = this.asset.scene;
+    this.addTouchDots();
+    this.findNodes();
+    this.add(this.rootNode);
+    this.loaded = true;
+  }
+
+  /**
+   * Polls data from the XRInputSource and updates the model's components to match
+   * the real world data
+   */
+  updateMatrixWorld(force) {
+    super.updateMatrixWorld(force);
+
+    if (!this.loaded) {
+      return;
+    }
+
+    // Cause the MotionController to poll the Gamepad for data
+    this.motionController.updateFromGamepad();
+
+    // Update the 3D model to reflect the button, thumbstick, and touchpad state
+    Object.values(this.motionController.components).forEach((component) => {
+      // Update node data based on the visual responses' current states
+      Object.values(component.visualResponses).forEach((visualResponse) => {
+        const {
+          valueNodeName, minNodeName, maxNodeName, value, valueNodeProperty
+        } = visualResponse;
+        const valueNode = this.nodes[valueNodeName];
+
+        // Skip if the visual response node is not found. No error is needed,
+        // because it will have been reported at load time.
+        if (!valueNode) return;
+
+        // Calculate the new properties based on the weight supplied
+        if (valueNodeProperty === Constants.VisualResponseProperty.VISIBILITY) {
+          valueNode.visible = value;
+        } else if (valueNodeProperty === Constants.VisualResponseProperty.TRANSFORM) {
+          const minNode = this.nodes[minNodeName];
+          const maxNode = this.nodes[maxNodeName];
+          THREE.Quaternion.slerp(
+            minNode.quaternion,
+            maxNode.quaternion,
+            valueNode.quaternion,
+            value
+          );
+
+          valueNode.position.lerpVectors(
+            minNode.position,
+            maxNode.position,
+            value
+          );
+        }
+      });
+    });
+  }
+
+  /**
+   * Walks the model's tree to find the nodes needed to animate the components and
+   * saves them for use in the frame loop
+   */
+  findNodes() {
+    this.nodes = {};
```

Comment:
You're setting `this.nodes` to a dictionary here, but initializing it to an empty array in the constructor.